// aai/esm2020/lib/core/custom-http-interceptor.mjs
import { HttpErrorResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { tap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "./auth-config.service";
export const interceptorFactory = (config) => {
    return new CustomHttpInterceptor(config);
};
export class CustomHttpInterceptor {
    constructor(authConfig) {
        this.authConfig = authConfig;
    }
    intercept(req, next) {
        let authReq = req;
        if (!(new RegExp(this.authConfig.authConfig.notAuthUrl.split(',').join('$|') + '$').test(req.url))) {
            // Get the auth header from the service.
            const authHeaderName = 'Authorization';
            // Clone the request to add the new header.
            authReq = req.clone({
                headers: req.headers.set(authHeaderName, this.authConfig.getAuthorizationHeader())
            });
        }
        // Pass on the cloned request instead of the original request.
        return next.handle(authReq).pipe(tap(() => {
            // if (event instanceof HttpResponse) {
            //     // do stuff with response if you want
            // }
        }, error => {
            // Remember, there may be other events besides just the response.
            if (error instanceof HttpErrorResponse) {
                if (error.status === 401) {
                    console.log('登陆过期');
                }
            }
            return;
        }));
    }
}
CustomHttpInterceptor.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: CustomHttpInterceptor, deps: [{ token: i1.AuthConfigService }], target: i0.ɵɵFactoryTarget.Injectable });
CustomHttpInterceptor.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: CustomHttpInterceptor });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: CustomHttpInterceptor, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.AuthConfigService }]; } });
