// aai/esm2020/lib/core/message.service.mjs
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class MessageModel {
}
var MsgType;
(function (MsgType) {
    MsgType["error"] = "\u9519\u8BEF";
    MsgType["success"] = "\u64CD\u4F5C\u6210\u529F";
    MsgType["info"] = "\u6CE8\u610F";
    MsgType["warning"] = "\u8B66\u544A";
})(MsgType || (MsgType = {}));
export class MessageService {
    constructor() {
        this.loadingSubject = new EventEmitter();
        this.growlSubject = new EventEmitter();
    }
    showLoading() {
        this.loadingSubject.emit(false);
    }
    hideLoading() {
        this.loadingSubject.emit(true);
    }
    success(detail, summary = 'success') {
        const content = {
            type: summary,
            title: MsgType[summary],
            content: detail
        };
        this.growlSubject.emit(content);
    }
    info(detail, summary = 'info') {
        const content = {
            type: summary,
            title: MsgType[summary],
            content: detail
        };
        this.growlSubject.emit(content);
    }
    warn(detail, summary = 'warning') {
        const content = {
            type: summary,
            title: MsgType[summary],
            content: detail
        };
        this.growlSubject.emit(content);
    }
    error(detail, summary = 'error') {
        const content = {
            type: summary,
            title: MsgType[summary],
            content: detail
        };
        this.growlSubject.emit(content);
    }
}
MessageService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: MessageService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
MessageService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: MessageService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: MessageService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
