// aai/esm2020/component/loading/loading.component.mjs
import { Component, Input } from '@angular/core';
import { NavigationCancel, NavigationEnd, NavigationError, NavigationStart } from '@angular/router';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../../lib/core/message.service";
export class LoadingComponent {
    constructor(router, msg) {
        this.router = router;
        this.msg = msg;
        this.hide = true;
        this.msg.loadingSubject.subscribe((v) => {
            this.hide = v;
        });
        this.router.events.subscribe((event) => {
            // show loading
            if (event instanceof NavigationStart) {
                this.hide = false;
            }
            if (event instanceof NavigationEnd ||
                event instanceof NavigationCancel ||
                event instanceof NavigationError) {
                setTimeout(() => {
                    this.hide = true;
                }, 100);
            }
        });
    }
}
LoadingComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: LoadingComponent, deps: [{ token: i1.Router }, { token: i2.MessageService }], target: i0.ɵɵFactoryTarget.Component });
LoadingComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.0.1", type: LoadingComponent, selector: "pz-loading", inputs: { hide: "hide" }, ngImport: i0, template: `
    <div class="spinner" [hidden]="hide">
      <div></div>
    </div>
  `, isInline: true, styles: ["@keyframes spin{0%{transform:rotate(0)}to{transform:rotate(360deg)}}.spinner{position:fixed;top:0;left:0;width:100%;height:100%;z-index:1003;background:rgba(0,0,0,.8);overflow:hidden}.spinner div:first-child{display:block;position:relative;left:50%;top:50%;width:150px;height:150px;margin:-75px 0 0 -75px;border-radius:50%;box-shadow:0 3px 3px #ff386a;transform:translateZ(0);animation:spin 2s linear infinite}.spinner div:first-child:after,.spinner div:first-child:before{content:\"\";position:absolute;border-radius:50%}.spinner div:first-child:before{inset:5px;box-shadow:0 3px 3px #ffe420;animation:spin 3s linear infinite}.spinner div:first-child:after{inset:15px;box-shadow:0 3px 3px #3dafff;animation:spin 1.5s linear infinite}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: LoadingComponent, decorators: [{
            type: Component,
            args: [{ selector: 'pz-loading', template: `
    <div class="spinner" [hidden]="hide">
      <div></div>
    </div>
  `, styles: ["@keyframes spin{0%{transform:rotate(0)}to{transform:rotate(360deg)}}.spinner{position:fixed;top:0;left:0;width:100%;height:100%;z-index:1003;background:rgba(0,0,0,.8);overflow:hidden}.spinner div:first-child{display:block;position:relative;left:50%;top:50%;width:150px;height:150px;margin:-75px 0 0 -75px;border-radius:50%;box-shadow:0 3px 3px #ff386a;transform:translateZ(0);animation:spin 2s linear infinite}.spinner div:first-child:after,.spinner div:first-child:before{content:\"\";position:absolute;border-radius:50%}.spinner div:first-child:before{inset:5px;box-shadow:0 3px 3px #ffe420;animation:spin 3s linear infinite}.spinner div:first-child:after{inset:15px;box-shadow:0 3px 3px #3dafff;animation:spin 1.5s linear infinite}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.MessageService }]; }, propDecorators: { hide: [{
                type: Input
            }] } });
