// aai/esm2020/component/qr-code/qr-code.component.mjs
import { Component, Input, ViewChild } from '@angular/core';
import * as QRCode from 'qrious';
import * as i0 from "@angular/core";
export class QrCodeComponent {
    constructor() {
        //
    }
    set content(content) {
        this.createQrcode(content);
    }
    ngOnInit() {
    }
    createQrcode(content) {
        const qr = new QRCode({
            element: this.ele.nativeElement,
            value: `${content}`
        });
        // 二维码的误差校正级别(L, M, Q, H)
        qr.level = 'M';
        qr.size = 70;
    }
}
QrCodeComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: QrCodeComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
QrCodeComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.0.1", type: QrCodeComponent, selector: "pz-qr-code", inputs: { content: "content" }, viewQueries: [{ propertyName: "ele", first: true, predicate: ["qr"], descendants: true, static: true }], ngImport: i0, template: "<canvas #qr></canvas>\n" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: QrCodeComponent, decorators: [{
            type: Component,
            args: [{ selector: 'pz-qr-code', template: "<canvas #qr></canvas>\n" }]
        }], ctorParameters: function () { return []; }, propDecorators: { content: [{
                type: Input
            }], ele: [{
                type: ViewChild,
                args: ['qr', { static: true }]
            }] } });
