// aai/esm2020/component/verify/verify.component.mjs
import { Component, EventEmitter, Output } from '@angular/core';
import * as i0 from "@angular/core";
function getRandomNumberByRange(start, end) {
    return Math.round(Math.random() * (end - start) + start);
}
function createCanvas(width, height) {
    const canvas = createElement('canvas');
    canvas.width = width;
    canvas.height = height;
    return canvas;
}
function createImg(onload) {
    const img = createElement('img');
    img.crossOrigin = 'Anonymous';
    img.onload = onload;
    img.onerror = () => {
        img.src = getRandomImg();
    };
    img.src = getRandomImg();
    return img;
}
function createElement(tagName) {
    return document.createElement(tagName);
}
function addClass(tag, className) {
    tag.classList.add(className);
}
function removeClass(tag, className) {
    tag.classList.remove(className);
}
function getRandomImg() {
    return 'https://picsum.photos/300/150/?image=' + getRandomNumberByRange(0, 100);
}
function sum(x, y) {
    return x + y;
}
function square(x) {
    return x * x;
}
export class VerifyComponent {
    constructor() {
        this.sideWidth = 300;
        this.sideHeight = 150;
        this.sideRadius = 10;
        this.sideLength = 42;
        this.submit = new EventEmitter();
        //
    }
    ngOnInit() {
        this.init();
    }
    draw(ctx, operation, x, y) {
        const l = this.sideLength;
        const PI = Math.PI;
        const r = this.sideRadius;
        ctx.beginPath();
        ctx.moveTo(x, y);
        ctx.lineTo(x + l / 2, y);
        ctx.arc(x + l / 2, y - r + 2, r, 0, 2 * PI);
        ctx.lineTo(x + l / 2, y);
        ctx.lineTo(x + l, y);
        ctx.lineTo(x + l, y + l / 2);
        ctx.arc(x + l + r - 2, y + l / 2, r, 0, 2 * PI);
        ctx.lineTo(x + l, y + l / 2);
        ctx.lineTo(x + l, y + l);
        ctx.lineTo(x, y + l);
        ctx.lineTo(x, y);
        ctx.fillStyle = '#fff';
        ctx[operation]();
        ctx.beginPath();
        ctx.arc(x, y + l / 2, r, 1.5 * PI, 0.5 * PI);
        ctx.globalCompositeOperation = 'xor';
        ctx.fill();
    }
    init() {
        this.reallyLength = this.sideLength + this.sideRadius * 2;
        this.root = document.getElementById('verify');
        this.initDOM();
        this.initImg();
        this.drawPic();
        this.bindEvents();
    }
    initDOM() {
        const canvas = createCanvas(this.sideWidth, this.sideHeight); // 画布
        const block = canvas.cloneNode(true); // 滑块
        const sliderContainer = createElement('div');
        const refreshIcon = createElement('div');
        const sliderMask = createElement('div');
        const slider = createElement('div');
        const sliderIcon = createElement('span');
        const text = createElement('span');
        block.className = 'block';
        sliderContainer.className = 'sliderContainer';
        refreshIcon.className = 'refreshIcon';
        refreshIcon.innerHTML = '<svg viewBox="64 64 896 896" fill="currentColor"\ ' +
            'width="1em" height="1em" data-icon="redo" aria-hidden="true">' +
            '<path d="M758.2 839.1C851.8 765.9 912 651.9 912 523.9 912 303 733.5' +
            ' 124.3 512.6 124 291.4 123.7 112 302.8 112 523.9c0 125.2 57.5 236.9' +
            ' 147.6 310.2 3.5 2.8 8.6 2.2 11.4-1.3l39.4-50.5c2.7-3.4' +
            ' 2.1-8.3-1.2-11.1-8.1-6.6-15.9-13.7-23.4-21.2a318.64 318.64' +
            ' 0 0 1-68.6-101.7C200.4 609 192 567.1 192 523.9s8.4-85.1' +
            ' 25.1-124.5c16.1-38.1 39.2-72.3 68.6-101.7 29.4-29.4 63.6-52.5' +
            ' 101.7-68.6C426.9 212.4 468.8 204 512 204s85.1 8.4 124.5' +
            ' 25.1c38.1 16.1 72.3 39.2 101.7 68.6 29.4 29.4 52.5 63.6' +
            ' 68.6 101.7 16.7 39.4 25.1 81.3 25.1 124.5s-8.4 85.1-25.1' +
            ' 124.5a318.64 318.64 0 0 1-68.6 101.7c-9.3 9.3-19.1 18-29.3' +
            ' 26L668.2 724a8 8 0 0 0-14.1 3l-39.6 162.2c-1.2 5 2.6 9.9 7.7' +
            ' 9.9l167 .8c6.7 0 10.5-7.7 6.3-12.9l-37.3-47.9z"></path>' +
            '</svg>';
        sliderMask.className = 'sliderMask';
        slider.className = 'slider';
        sliderIcon.className = 'sliderIcon';
        sliderIcon.innerHTML = '<svg viewBox="64 64 896 896" fill="currentColor" width="1em" ' +
            'height="1em" data-icon="arrow-right" aria-hidden="true"><path d="M869 487.8L491.2 ' +
            '159.9c-2.9-2.5-6.6-3.9-10.5-3.9h-88.5c-7.4 0-10.8 9.2-5.2 14l350.2 304H152c-4.4 0-8 ' +
            '3.6-8 8v60c0 4.4 3.6 8 8 8h585.1L386.9 854c-5.6 4.9-2.2 14 5.2 14h91.5c1.9 0 3.8-.7 ' +
            '5.2-2L869 536.2a32.07 32.07 0 0 0 0-48.4z"></path></svg>';
        text.innerHTML = '<svg viewBox="0 0 300 40">' +
            '  <linearGradient id="gr-radial"  x1="0%" y1="0%" x2="100%" y2="0%">' +
            '      <stop stop-color="#333" offset="0%"/>' +
            '      <stop stop-opacity="0.1" offset="50%">' +
            '          <animate attributeName="offset" from="0" to="1.3"' +
            '           dur="3s" repeatCount="indefinite" />' +
            '      </stop>' +
            '      <stop stop-color="#333" offset="100%"/>' +
            '  </linearGradient>' +
            '  <text text-anchor="middle" x="50%" y="50%" dy="5px"' +
            '  style="font-size:12pt; fill: url(#gr-radial)">向右滑动滑块填充拼图</text>' +
            '</svg>';
        text.className = 'sliderText';
        const el = this.root;
        el.appendChild(canvas);
        el.appendChild(refreshIcon);
        el.appendChild(block);
        slider.appendChild(sliderIcon);
        sliderMask.appendChild(slider);
        sliderContainer.appendChild(sliderMask);
        sliderContainer.appendChild(text);
        el.appendChild(sliderContainer);
        Object.assign(this, {
            canvas,
            block,
            sliderContainer,
            refreshIcon,
            slider,
            sliderMask,
            sliderIcon,
            text,
            canvasCtx: canvas.getContext('2d'),
            blockCtx: block.getContext('2d')
        });
    }
    initImg() {
        const img = createImg(() => {
            this.canvasCtx.drawImage(img, 0, 0, this.sideWidth, this.sideHeight);
            this.blockCtx.drawImage(img, 0, 0, this.sideWidth, this.sideHeight);
            const y = this.yPos - this.sideRadius * 2 + 2;
            const imageData = this.blockCtx.getImageData(this.xPos, y, this.reallyLength, this.reallyLength);
            this.block.width = this.reallyLength;
            this.blockCtx.putImageData(imageData, 0, y);
        });
        this.img = img;
    }
    drawPic() {
        // 随机创建滑块的位置
        this.xPos = getRandomNumberByRange(this.reallyLength + 10, this.sideWidth - (this.reallyLength + 10));
        this.yPos = getRandomNumberByRange(10 + this.sideRadius * 2, this.sideHeight - (this.reallyLength + 10));
        this.draw(this.canvasCtx, 'fill', this.xPos, this.yPos);
        this.draw(this.blockCtx, 'clip', this.xPos, this.yPos);
    }
    clean() {
        this.canvasCtx.clearRect(0, 0, this.sideWidth, this.sideHeight);
        this.blockCtx.clearRect(0, 0, this.sideWidth, this.sideHeight);
        this.block.width = this.sideWidth;
    }
    bindEvents() {
        this.root.onselectstart = () => false;
        this.refreshIcon.onclick = () => {
            this.reset();
        };
        let originX, originY, isMouseDown = false;
        const trail = [];
        this.slider.addEventListener('mousedown', function (e) {
            originX = e.x;
            originY = e.y;
            isMouseDown = true;
        });
        document.addEventListener('mousemove', (e) => {
            if (!isMouseDown) {
                return false;
            }
            const moveX = e.x - originX;
            const moveY = e.y - originY;
            if (moveX < 0 || moveX + 38 >= this.sideWidth) {
                return false;
            }
            this.slider.style.left = moveX + 'px';
            const blockLeft = (this.sideWidth - 40 - 20) / (this.sideWidth - 40) * moveX;
            this.block.style.left = blockLeft + 'px';
            if (moveX) {
                addClass(this.sliderContainer, 'sliderContainer_active');
            }
            else {
                removeClass(this.sliderContainer, 'sliderContainer_active');
            }
            this.sliderMask.style.width = (moveX ? (moveX + 15) : moveX) + 'px';
            trail.push(moveY);
        });
        //
        document.addEventListener('mouseup', (e) => {
            if (!isMouseDown) {
                return false;
            }
            isMouseDown = false;
            if (e.x === originX) {
                return false;
            }
            removeClass(this.sliderContainer, 'sliderContainer_active');
            this.trail = trail;
            const { spliced, TuringTest } = this.verify();
            if (spliced) {
                if (TuringTest) {
                    addClass(this.sliderContainer, 'sliderContainer_success');
                    this.success();
                }
                else {
                    addClass(this.sliderContainer, 'sliderContainer_fail');
                    this.text.innerHTML = '再试一次';
                    this.reset();
                }
            }
            else {
                if (e.x - originX > 0) {
                    addClass(this.sliderContainer, 'sliderContainer_fail');
                    this.fail();
                    setTimeout(() => {
                        this.reset();
                    }, 1000);
                }
            }
        });
    }
    verify() {
        const arr = this.trail; // 拖动时y轴的移动距离
        const average = arr.reduce(sum) / arr.length; // 平均值
        const deviations = arr.map(x => x - average); // 偏差数组
        const stddev = Math.sqrt(deviations.map(square).reduce(sum) / arr.length); // 标准差
        const left = parseInt(this.block.style.left);
        return {
            spliced: Math.abs(left - this.xPos) < 10,
            TuringTest: average !== stddev, // 只是简单的验证拖动轨迹，相等时一般为0，表示可能非人为操作
        };
    }
    reset() {
        this.sliderContainer.className = 'sliderContainer';
        this.slider.style.left = 0;
        this.block.style.left = 0;
        this.sliderMask.style.width = 0;
        this.clean();
        this.img.src = getRandomImg();
        this.drawPic();
    }
    success() {
        this.submit.emit(true);
        setTimeout(() => this.reset(), 1000);
    }
    fail() {
        this.submit.emit(false);
    }
}
VerifyComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: VerifyComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
VerifyComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.0.1", type: VerifyComponent, selector: "pz-verify", outputs: { submit: "submit" }, ngImport: i0, template: "<div id=\"verify\"></div>\n", styles: ["::ng-deep #verify{position:relative;width:300px}::ng-deep #verify canvas:first-child{border-radius:8px;background:rgba(255,255,255,.2)}::ng-deep #verify .block{position:absolute;left:0;top:0}::ng-deep #verify .sliderContainer{position:relative;text-align:center;width:300px;height:40px;line-height:40px;background:#f7f9fa;color:#45494c;border:1px solid #e4e7eb;border-radius:5px}::ng-deep #verify .sliderContainer_active .slider{top:-1px;height:38px}::ng-deep #verify .sliderContainer_active .sliderMask{height:38px;border-width:1px}::ng-deep #verify .sliderContainer_success .slider{height:38px;top:-1px;border:1px solid #52CCBA;background-color:#52ccba!important}::ng-deep #verify .sliderContainer_success .sliderMask{height:38px;border:1px solid #52CCBA;background-color:#d2f4ef}::ng-deep #verify .sliderContainer_success .sliderIcon{background-position:0 0!important}::ng-deep #verify .sliderContainer_fail .slider{top:-1px;height:38px;background-color:#f57a7a!important}::ng-deep #verify .sliderContainer_fail .sliderMask{height:38px;border:1px solid #f57a7a;background-color:#fce1e1}::ng-deep #verify .sliderContainer_fail .sliderIcon{background-position:0 -83px!important}::ng-deep #verify .sliderContainer_active .sliderText,::ng-deep #verify .sliderContainer_success .sliderText,::ng-deep #verify .sliderContainer_fail .sliderText{display:none}::ng-deep #verify .sliderMask{position:absolute;left:0;top:0;height:39px;border:0 solid #1991FA;background:#D1E9FE;border-radius:5px}::ng-deep #verify .slider{position:absolute;top:0;left:0;width:38px;height:38px;background:#fff;box-shadow:0 0 3px #0000004d;cursor:pointer;transition:background .2s linear;border-radius:5px}::ng-deep #verify .slider:hover{background:#1991FA}::ng-deep #verify .slider:hover .sliderIcon{color:#fff}::ng-deep #verify .sliderIcon{position:absolute;top:2px;left:12px;width:18px;height:18px}::ng-deep #verify .sliderIcon>svg{height:100%;width:100%}::ng-deep #verify .refreshIcon{color:#fff;position:absolute;right:0;top:0;width:25px;height:25px;cursor:pointer}::ng-deep #verify .refreshIcon>svg{height:90%;width:90%;transform:rotate(-180deg);opacity:.7}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: VerifyComponent, decorators: [{
            type: Component,
            args: [{ selector: 'pz-verify', template: "<div id=\"verify\"></div>\n", styles: ["::ng-deep #verify{position:relative;width:300px}::ng-deep #verify canvas:first-child{border-radius:8px;background:rgba(255,255,255,.2)}::ng-deep #verify .block{position:absolute;left:0;top:0}::ng-deep #verify .sliderContainer{position:relative;text-align:center;width:300px;height:40px;line-height:40px;background:#f7f9fa;color:#45494c;border:1px solid #e4e7eb;border-radius:5px}::ng-deep #verify .sliderContainer_active .slider{top:-1px;height:38px}::ng-deep #verify .sliderContainer_active .sliderMask{height:38px;border-width:1px}::ng-deep #verify .sliderContainer_success .slider{height:38px;top:-1px;border:1px solid #52CCBA;background-color:#52ccba!important}::ng-deep #verify .sliderContainer_success .sliderMask{height:38px;border:1px solid #52CCBA;background-color:#d2f4ef}::ng-deep #verify .sliderContainer_success .sliderIcon{background-position:0 0!important}::ng-deep #verify .sliderContainer_fail .slider{top:-1px;height:38px;background-color:#f57a7a!important}::ng-deep #verify .sliderContainer_fail .sliderMask{height:38px;border:1px solid #f57a7a;background-color:#fce1e1}::ng-deep #verify .sliderContainer_fail .sliderIcon{background-position:0 -83px!important}::ng-deep #verify .sliderContainer_active .sliderText,::ng-deep #verify .sliderContainer_success .sliderText,::ng-deep #verify .sliderContainer_fail .sliderText{display:none}::ng-deep #verify .sliderMask{position:absolute;left:0;top:0;height:39px;border:0 solid #1991FA;background:#D1E9FE;border-radius:5px}::ng-deep #verify .slider{position:absolute;top:0;left:0;width:38px;height:38px;background:#fff;box-shadow:0 0 3px #0000004d;cursor:pointer;transition:background .2s linear;border-radius:5px}::ng-deep #verify .slider:hover{background:#1991FA}::ng-deep #verify .slider:hover .sliderIcon{color:#fff}::ng-deep #verify .sliderIcon{position:absolute;top:2px;left:12px;width:18px;height:18px}::ng-deep #verify .sliderIcon>svg{height:100%;width:100%}::ng-deep #verify .refreshIcon{color:#fff;position:absolute;right:0;top:0;width:25px;height:25px;cursor:pointer}::ng-deep #verify .refreshIcon>svg{height:90%;width:90%;transform:rotate(-180deg);opacity:.7}\n"] }]
        }], ctorParameters: function () { return []; }, propDecorators: { submit: [{
                type: Output
            }] } });
